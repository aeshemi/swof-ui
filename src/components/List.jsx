import React, { Component } from 'react';
import { Button, Col, Glyphicon, Grid, Row } from 'react-bootstrap';
import moment from 'moment';
import ListRow from './ListRow';
import ApiClient from '../ApiClient';

class List extends Component {
  constructor() {
    super();
    this.state = { currentDate: moment(), schedule: [] };
    this.getSchedule = this.getSchedule.bind(this);
    this.updateSchedule = this.updateSchedule.bind(this);
  }

  componentDidMount() {
    this.getSchedule(this.state.currentDate);
  }

  getSchedule(date) {
    ApiClient.getList(date.format('YYYY-MM-DD')).then(data => {
      const schedule = Object.values(
        data.reduce((s, d) => {
          /* eslint-disable no-param-reassign */
          const startDate = moment(d.date);

          s[d.date] = s[d.date] || {
            day: startDate.date(),
            dayOfWeek: startDate.format('ddd'),
            engineer: {}
          };

          s[d.date].engineer[d.shift] = d.name;

          return s;
          /* eslint-enable no-param-reassign */
        }, {})
      );

      this.setState({ schedule });
    });
  }

  updateSchedule(isForward) {
    const { currentDate } = this.state;
    const startDate = isForward ? currentDate.add(1, 'month') : currentDate.subtract(1, 'month');
    this.getSchedule(startDate.startOf('month'));
  }

  render() {
    const { currentDate, schedule } = this.state;

    const noRotations = <div className="no-rotations">No rotations to display.</div>;

    return (
      <div className="schedule-list">
        <Grid>
          <Row>
            <Col md={3}>
              <Button onClick={() => this.updateSchedule(false)}>
                <Glyphicon glyph="chevron-left" />
              </Button>
            </Col>
            <Col md={6} className="list-month">
              {currentDate.format('MMMM YYYY')}
            </Col>
            <Col md={3}>
              <Button className="pull-right" onClick={() => this.updateSchedule(true)}>
                <Glyphicon glyph="chevron-right" />
              </Button>
            </Col>
          </Row>
        </Grid>
        {schedule.length === 0 ? noRotations : schedule.map(s => <ListRow key={s.day} {...s} />)}
      </div>
    );
  }
}

export default List;
