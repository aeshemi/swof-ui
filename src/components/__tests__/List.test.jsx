import List from '../List';

describe('List', () => {
  beforeEach(() => {
    fetch.mockResponse(JSON.stringify([]));
  });

  afterEach(() => {
    fetch.resetMocks();
  });

  it('should call getSchedule during componentDidMount', () => {
    const spy = jest.spyOn(List.prototype, 'getSchedule');

    mount(<List />);

    expect(spy).toHaveBeenCalled();

    spy.mockReset();
    spy.mockRestore();
  });

  it('renders correctly with no schedules', done => {
    const wrapper = shallow(<List />);

    setImmediate(() => {
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  it('renders correctly with schedules', done => {
    fetch.mockResponse(
      JSON.stringify([
        { date: '2018-08-20T00:00:00', shift: 'Morning', name: 'Hetti Whatley' },
        { date: '2018-08-20T00:00:00', shift: 'Afternoon', name: 'Natalee Wigin' },
        { date: '2018-08-21T00:00:00', shift: 'Morning', name: 'Charlot Eede' },
        { date: '2018-08-21T00:00:00', shift: 'Afternoon', name: 'Monti Vercruysse' },
        { date: '2018-08-22T00:00:00', shift: 'Morning', name: 'Nicolea Appleyard' },
        { date: '2018-08-22T00:00:00', shift: 'Afternoon', name: 'Nap Nimmo' },
        { date: '2018-08-23T00:00:00', shift: 'Morning', name: 'Lethia Bonefant' },
        { date: '2018-08-23T00:00:00', shift: 'Afternoon', name: 'Sharity Dronsfield' },
        { date: '2018-08-24T00:00:00', shift: 'Morning', name: 'Auria De Witt' },
        { date: '2018-08-24T00:00:00', shift: 'Afternoon', name: 'Burk Maciaszek' }
      ])
    );

    const wrapper = shallow(<List />);

    setImmediate(() => {
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  it('click on previous/forward calls updateSchedule', () => {
    const spy = jest.spyOn(List.prototype, 'updateSchedule');
    const wrapper = shallow(<List />);

    wrapper
      .find('Button')
      .first()
      .simulate('click');

    expect(spy).toHaveBeenCalledWith(false);

    wrapper
      .find('Button')
      .last()
      .simulate('click');

    expect(spy).toHaveBeenCalledWith(true);

    spy.mockReset();
    spy.mockRestore();
  });
});
