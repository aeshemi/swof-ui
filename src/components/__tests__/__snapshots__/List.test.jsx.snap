// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders correctly with no schedules 1`] = `
<div
  className="schedule-list"
>
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        md={3}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          disabled={false}
          onClick={[Function]}
        >
          <Glyphicon
            bsClass="glyphicon"
            glyph="chevron-left"
          />
        </Button>
      </Col>
      <Col
        bsClass="col"
        className="list-month"
        componentClass="div"
        md={6}
      >
        August 2018
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        md={3}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="pull-right"
          disabled={false}
          onClick={[Function]}
        >
          <Glyphicon
            bsClass="glyphicon"
            glyph="chevron-right"
          />
        </Button>
      </Col>
    </Row>
  </Grid>
  <div
    className="no-rotations"
  >
    No rotations to display.
  </div>
</div>
`;

exports[`List renders correctly with schedules 1`] = `
<div
  className="schedule-list"
>
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        md={3}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          disabled={false}
          onClick={[Function]}
        >
          <Glyphicon
            bsClass="glyphicon"
            glyph="chevron-left"
          />
        </Button>
      </Col>
      <Col
        bsClass="col"
        className="list-month"
        componentClass="div"
        md={6}
      >
        August 2018
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        md={3}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="pull-right"
          disabled={false}
          onClick={[Function]}
        >
          <Glyphicon
            bsClass="glyphicon"
            glyph="chevron-right"
          />
        </Button>
      </Col>
    </Row>
  </Grid>
  <ListRow
    day={20}
    dayOfWeek="Mon"
    engineer={
      Object {
        "Afternoon": "Natalee Wigin",
        "Morning": "Hetti Whatley",
      }
    }
    key="20"
  />
  <ListRow
    day={21}
    dayOfWeek="Tue"
    engineer={
      Object {
        "Afternoon": "Monti Vercruysse",
        "Morning": "Charlot Eede",
      }
    }
    key="21"
  />
  <ListRow
    day={22}
    dayOfWeek="Wed"
    engineer={
      Object {
        "Afternoon": "Nap Nimmo",
        "Morning": "Nicolea Appleyard",
      }
    }
    key="22"
  />
  <ListRow
    day={23}
    dayOfWeek="Thu"
    engineer={
      Object {
        "Afternoon": "Sharity Dronsfield",
        "Morning": "Lethia Bonefant",
      }
    }
    key="23"
  />
  <ListRow
    day={24}
    dayOfWeek="Fri"
    engineer={
      Object {
        "Afternoon": "Burk Maciaszek",
        "Morning": "Auria De Witt",
      }
    }
    key="24"
  />
</div>
`;
